pymc_experimental/tests/custom_rjmcmc_sketch.py

```diff
@@ -0,0 +1,333 @@
+    values - the gradients in each dimension.
+    """
+
+    itypes = [tt.dvector]
+    otypes = [tt.dvector]
+
+    def __init__(self, loglike):
+        """
+        Initialise with various things that the function requires. Below
+        are the things that are needed in this particular example.
+
+        Parameters
+        ----------
+        loglike:
+            The log-likelihood (or whatever) function we've defined
+        data:
+            The "observed" data that our log-likelihood function takes in
+        x:
+            The dependent variable (aka 'x') that our model requires
+        sigma:
+            The noise standard deviation that out function requires.
+        """
+
+        # add inputs as class attributes
+        self.likelihood = loglike
+
+    def perform(self, node, inputs, outputs):
+        (theta,) = inputs
+
+        outputs[0][0] = scipy.optimize.approx_fprime(theta, self.likelihood, epsilon= np.sqrt(np.finfo(float).eps))
+
+
+
+# Generate Data from the true model
+timepoints = np.linspace(0, 5, 1000)
+true_deltas = np.array([[0, 1],[0, 0]])
+true_ks     = np.array([[0, 2],[0, 0]])
+initial_values = np.array([5, 0])
+true_model = get_vectorized_solution_for_matrix(true_deltas,true_ks, initial_values)
+true_data = [true_model[i](timepoints) for i in range(2)]
+
+# Define loglikelihood 
+def my_loglike(vec):
+    # we'll just hardcode this cause I'm tired
+
+    if (vec[0] != 0 and vec[0] != 1) and (vec[1] != 0 and vec[1] != 1):
+        raise(ValueError('Delta parameters can only be one or zero'))
+
+    deltas = np.array([[0, vec[0]],[vec[1], 0]])
+    ks = np.array([[0, vec[2]],[vec[3], 0]])
+    new_model = get_vectorized_solution_for_matrix(deltas, ks, initial_values)
+    simulated_data = [new_model[i](timepoints) for i in range(2)]
+    sum = 0
+    for i in range(2):
+        sum += np.sum(np.log(scipy.stats.norm.pdf(simulated_data[i] - true_data[i])))
+
+    if not np.isfinite(sum):
+        raise(ValueError('loglikelihood diverged'))
+
+    return sum
+
+# Generate the likelihood
+ndraws = 200
+nburn = 50
+pymc_model = pm.Model()
+theano_loglike = LogLike(lambda x: my_loglike(x))
+with pymc_model:
+    # Generate priors based on the Supermodel priors
+    var_assoc = []
+    delta_parameters = []
+    k_parameters = []
+    for i in range(2):
+        for j in range(2):
+            if i!=j:
+                delta = pm.Bernoulli('delta_{}_{}'.format(i, j), 0.5)
+                k = pm.Uniform('k_{}_{}'.format(i, j), lower=0, upper=10)
+                var_assoc.append((delta, k))
+                delta_parameters.append(delta)
+                k_parameters.append(k)
+    
+    pymc_parameters = [d for d, _ in var_assoc] + [k for _, k in var_assoc]
+
+    theta = tt.as_tensor_variable(pymc_parameters)
+
+    pm.Potential("loglikelihood", theano_loglike(theta))
+
+    
+
+    class Jump(pm.step_methods.arraystep.BlockedStep):
```

Comment:
Any class definition should be independent of the model (0th indentation level).

If the instance needs access to the `pymc_model`, it can get that via `pm.modelcontext()`, because the stepper will be instantiated inside the activated modelcontext.

If the instance needs references to other variables (e.g. `k_parameters`), they must be passed to the `__init__` and maybe kept as instance attributes of the stepper.

Reply:
indeed the class definition was a little too messy, I was still ideation phase in that local context.